src/python/zquantum/core/interfaces/backend_test.py

```diff
@@ -175,3 +374,97 @@ def test_get_bitstring_distribution_wf_simulators(self, wf_simulator):
             0.5, abs=1e-7
         )
+
+
+class QuantumSimulatorGatesTest:
+    @pytest.mark.parametrize(
+        "initial_gate,tested_gate,target_amplitudes",
+        one_qubit_non_parametric_gates_amplitudes_test_set,
+    )
+    def test_one_qubit_non_parametric_gates_using_amplitudes(
+        self, wf_simulator, initial_gate, tested_gate, target_amplitudes
+    ):
+        # Given
+        qubit_list = [Qubit(0)]
+        gate_1 = Gate(initial_gate, qubits=qubit_list)
+        gate_2 = Gate(tested_gate, qubits=qubit_list)
+
+        circuit = Circuit()
+        circuit.gates = [gate_1, gate_2]
+        import pdb
+
+        pdb.set_trace()
+
+        # When
+        wavefunction = wf_simulator.get_wavefunction(circuit)
+
+        # Then
+        assert np.allclose(wavefunction.amplitudes, target_amplitudes)
+
+    @pytest.mark.parametrize(
+        "initial_gate,tested_gate,params,target_amplitudes",
+        one_qubit_parametric_gates_amplitudes_test_set,
+    )
+    def test_one_qubit_parametric_gates_using_amplitudes(
+        self, wf_simulator, initial_gate, tested_gate, params, target_amplitudes
+    ):
+        # Given
+        qubit_list = [Qubit(0)]
+        gate_1 = Gate(initial_gate, qubits=qubit_list)
+        gate_2 = Gate(tested_gate, params=params, qubits=qubit_list)
+
+        circuit = Circuit()
+        circuit.gates = [gate_1, gate_2]
+
+        # When
+        wavefunction = wf_simulator.get_wavefunction(circuit)
+
+        # Then
+        assert np.allclose(wavefunction.amplitudes, target_amplitudes)
+
+    @pytest.mark.parametrize(
+        "initial_gates,tested_gate,target_amplitudes",
+        two_qubit_non_parametric_gates_amplitudes_test_set,
+    )
+    def test_two_qubit_non_parametric_gates_using_amplitudes(
+        self, wf_simulator, initial_gates, tested_gate, target_amplitudes
+    ):
+        # Given
+        qubit_list = [Qubit(0), Qubit(1)]
+        gate_1 = Gate(initial_gates[0], qubits=[qubit_list[0]])
+        gate_2 = Gate(initial_gates[1], qubits=[qubit_list[1]])
+        gate_3 = Gate(tested_gate, qubits=qubit_list)
+
+        circuit = Circuit()
+        circuit.gates = [gate_1, gate_2, gate_3]
+
+        # When
+        wavefunction = wf_simulator.get_wavefunction(circuit)
+
+        # Then
+        assert np.allclose(wavefunction.amplitudes, target_amplitudes)
+
+    @pytest.mark.parametrize(
+        "initial_gates,tested_gate,params,target_amplitudes",
+        two_qubit_parametric_gates_amplitudes_test_set,
+    )
+    def test_two_qubit_parametric_gates_using_amplitudes(
+        self, wf_simulator, initial_gates, tested_gate, params, target_amplitudes
+    ):
+        # Given
+        qubit_list = [Qubit(0), Qubit(1)]
+        gate_1 = Gate(initial_gates[0], qubits=[qubit_list[0]])
+        if initial_gates[1] == "CNOT":
+            gate_2 = Gate(initial_gates[1], qubits=qubit_list)
+        else:
+            gate_2 = Gate(initial_gates[1], qubits=[qubit_list[1]])
```

Comment:
Is only `initial_gates[1]` ever set to CNOT? No possibility for `initial_gates[0]`? Do we also make sure `len(initial_gates) == 2` always?

Reply:
Good catch, that's an artifact from some old code.

Reply:
Fixed in [bdab82](https://github.com/zapatacomputing/z-quantum-core/pull/145/commits/bdab8249a2b4f5b808b21b0ac7101368b29146a2)